internal/core/src/indexbuilder/KmeansMajorCompaction.cpp

```diff
@@ -0,0 +1,351 @@
+// limitations under the License.
+
+#include "index/VectorDiskIndex.h"
+
+#include "common/Tracer.h"
+#include "common/Types.h"
+#include "common/Utils.h"
+#include "config/ConfigKnowhere.h"
+#include "index/Meta.h"
+#include "index/Utils.h"
+#include "knowhere/kmeans.h"
+#include "indexbuilder/KmeansMajorCompaction.h"
+#include "indexbuilder/MajorCompaction.h"
+#include "segcore/SegcoreConfig.h"
+#include "storage/LocalChunkManagerSingleton.h"
+#include "storage/Util.h"
+#include "common/Consts.h"
+#include "common/RangeSearchHelper.h"
+#include "indexbuilder/types.h"
+#include <random>
+
+namespace milvus::indexbuilder {
+
+template <typename T>
+KmeansMajorCompaction<T>::KmeansMajorCompaction(
+    Config& config, const storage::FileManagerContext& file_manager_context)
+    : config_(config) {
+    file_manager_ =
+        std::make_shared<storage::DiskFileManagerImpl>(file_manager_context);
+    AssertInfo(file_manager_ != nullptr, "create file manager failed!");
+    auto local_chunk_manager =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    auto local_compaction_raw_data_path_prefix =
+        file_manager_->GetCompactionRawDataObjectPrefix();
+
+    if (local_chunk_manager->Exist(local_compaction_raw_data_path_prefix)) {
+        LOG_INFO("path {} already exists, delete it",
+                 local_compaction_raw_data_path_prefix);
+        local_chunk_manager->RemoveDir(local_compaction_raw_data_path_prefix);
+    }
+    local_chunk_manager->CreateDir(local_compaction_raw_data_path_prefix);
+}
+
+template <typename T>
+std::unique_ptr<T[]>
+KmeansMajorCompaction<T>::Sample(const std::vector<std::string>& file_paths,
+                                 const std::vector<uint64_t>& file_sizes,
+                                 uint64_t train_size,
+                                 uint64_t total_size) {
+    auto local_chunk_manager =
+        storage::LocalChunkManagerSingleton::GetInstance().GetChunkManager();
+    // train data fits in memory, read by sequence and generate centroids and id_mapping in one pass
+    if (train_size >= total_size) {
+        auto buf = std::unique_ptr<T[]>(new T[total_size / sizeof(T)]);
+        int64_t offset = 0;
+        for (int i = 0; i < file_paths.size(); i++) {
+            local_chunk_manager->Read(
+                file_paths[i],
+                0,
+                reinterpret_cast<char*>(buf.get()) + offset,
+                file_sizes[i]);
+            offset += file_sizes[i];
+        }
+        return buf;
+    }
+    // random sampling to get train data
+    std::vector<int32_t> idx(file_paths.size());
+    std::iota(idx.begin(), idx.end(), 0);
+    std::shuffle(idx.begin(), idx.end(), std::mt19937());
+    int selected_size = 0;
+    auto buf = std::unique_ptr<T[]>(new T[train_size / sizeof(T)]);
+    int64_t offset = 0;
+
+    for (int i = 0; i < file_paths.size(); i++) {
+        if (selected_size < train_size &&
```

Comment:
the logic here is correct, but very confusing

for (int i = 0; i < file_paths.size(); i++) {
    auto cur_size = std::min(train_size - selected_size, file_sizes[idx[i]]);
    local_chunk_manager->Read(
        file_paths[idx[i]], 0, reinterpret_cast<char*>(buf.get()) + offset, cur_size);
    selected_size += cur_size;
    offset += cur_size;

    if (selected_size >= train_size) {
        break;
    }
}